File: components/mardown-display/blocks/flashcards/flashcard-parser.ts
/**
 * Parser for flashcard content
 * Handles both "Front/Back" and "Question/Answer" formats
 * Supports optional thematic breaks (---)
 */

export interface Flashcard {
    front: string;
    back: string;
}

export interface FlashcardParseResult {
    flashcards: Flashcard[];
    isComplete: boolean;
    partialCard: Partial<Flashcard> | null;
}

/**
 * Parse flashcard content into individual cards
 * Only returns complete flashcards (both front and back present)
 * 
 * Supports both single-line and multi-line Back/Answer content:
 *   Back: single line answer
 *   Back: 
 *   - bullet one
 *   - bullet two
 */
export const parseFlashcards = (content: string): FlashcardParseResult => {
    const lines = content.split('\n');
    const flashcards: Flashcard[] = [];
    let currentCard: Partial<Flashcard> = {};
    let partialCard: Partial<Flashcard> | null = null;
    let isComplete = false;
    let collectingBack = false;
    let backLines: string[] = [];

    isComplete = content.includes('</flashcards>');

    const finalizeCard = () => {
        if (collectingBack && backLines.length > 0) {
            currentCard.back = backLines.join('\n').trim();
            collectingBack = false;
            backLines = [];
        }
        if (currentCard.front && currentCard.back) {
            flashcards.push({
                front: currentCard.front,
                back: currentCard.back,
            });
            currentCard = {};
        }
    };

    for (let i = 0; i < lines.length; i++) {
        const line = lines[i].trim();

        if (line === '---') {
            finalizeCard();
            continue;
        }

        const frontMatch = line.match(/^(?:Front|Question):\s*(.*)/i);
        if (frontMatch) {
            finalizeCard();
            currentCard.front = frontMatch[1].trim();
            collectingBack = false;
            backLines = [];
            continue;
        }

        const backMatch = line.match(/^(?:Back|Answer):\s*(.*)/i);
        if (backMatch) {
            collectingBack = true;
            backLines = [];
            const inlineContent = backMatch[1].trim();
            if (inlineContent) {
                backLines.push(inlineContent);
            }
            continue;
        }

        if (collectingBack) {
            if (line === '') {
                // Blank lines within back content are preserved
                backLines.push('');
            } else {
                backLines.push(line);
            }
            continue;
        }

        // Continuation of front text (no Back: seen yet for this card)
        if (currentCard.front && !currentCard.back && line) {
            currentCard.front += ' ' + line;
        }
    }

    // Finalize any in-progress back collection
    if (collectingBack && backLines.length > 0) {
        currentCard.back = backLines.join('\n').trim();
        collectingBack = false;
        backLines = [];
    }

    // Handle the last card
    if (currentCard.front && currentCard.back) {
        if (isComplete) {
            flashcards.push({
                front: currentCard.front,
                back: currentCard.back,
            });
            partialCard = null;
        } else {
            partialCard = currentCard;
        }
    } else if (currentCard.front || currentCard.back) {
        partialCard = currentCard;
    }

    return {
        flashcards,
        isComplete,
        partialCard,
    };
};

File: lib/python/parsers/flashcard_parser.py
"""Flashcard parser — port of flashcard-parser.ts."""

from __future__ import annotations

import re

from lib.python.models.flashcards import FlashcardItem, FlashcardsBlockData

_FRONT_RE = re.compile(r'^(?:Front|Question):\s*(.*)', re.IGNORECASE)
_BACK_RE = re.compile(r'^(?:Back|Answer):\s*(.*)', re.IGNORECASE)


def parse_flashcards(content: str) -> FlashcardsBlockData:
    """Parse flashcard content into structured data."""
    lines = content.split("\n")
    cards: list[FlashcardItem] = []
    current_front: str | None = None
    current_back_lines: list[str] = []
    partial_card: FlashcardItem | None = None
    collecting_back = False
    is_complete = "</flashcards>" in content

    def finalize_card() -> None:
        nonlocal current_front, current_back_lines, collecting_back
        if collecting_back and current_back_lines:
            back_text = "\n".join(current_back_lines).strip()
            if current_front and back_text:
                cards.append(FlashcardItem(front=current_front, back=back_text))
                current_front = None
            collecting_back = False
            current_back_lines = []

    for line in lines:
        stripped = line.strip()

        if stripped == "---":
            finalize_card()
            continue

        front_m = _FRONT_RE.match(stripped)
        if front_m:
            finalize_card()
            current_front = front_m.group(1).strip()
            collecting_back = False
            current_back_lines = []
            continue

        back_m = _BACK_RE.match(stripped)
        if back_m:
            collecting_back = True
            current_back_lines = []
            inline = back_m.group(1).strip()
            if inline:
                current_back_lines.append(inline)
            continue

        if collecting_back:
            current_back_lines.append(stripped if stripped else "")
            continue

        # Continuation of front text
        if current_front and not collecting_back and stripped:
            current_front += " " + stripped

    # Finalize in-progress back collection
    if collecting_back and current_back_lines:
        back_text = "\n".join(current_back_lines).strip()
        if current_front and back_text:
            if is_complete:
                cards.append(FlashcardItem(front=current_front, back=back_text))
            else:
                partial_card = FlashcardItem(front=current_front, back=back_text)
        elif current_front:
            partial_card = FlashcardItem(front=current_front, back="")
    elif current_front:
        partial_card = FlashcardItem(front=current_front, back="")

    return FlashcardsBlockData(cards=cards, is_complete=is_complete, partial_card=partial_card)
